### PR TITLE
crush: remove invalid upmap items

### PR DESCRIPTION
**Pull request by hjwsm1989**
We can not cancel in verify_upmap if remap an osd to different root bucket,

cluster topology:
osd.0 ~ osd.29 belongs to datacenter 1
osd30 ~ osd.59 belongs to datacenter 2
crush rule:
  take datacenter 1
  chooseleaf 2 host
  emit
  take datacenter 2
  chooseleaf 2 host
  emit
The pg's primary osd in datacenter 1.

We should cancel the pgid from upmap_items like below
1) from [26,12,54,46] to [30,12,54,46]
pg_upmap_items [26,30]

2) from [16,25,53,31] to [40,25,53,0]
pg_upmap_items [16,20,31,0]

Signed-off-by: huangjun <huangjun@xsky.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

**Comment by hjwsm1989** [comment]
@liewegas @xiexingguo please review this, thanks!

File: src/crush/CrushWrapper.cc [hunk 1]
```diff
@@ -952,13 +963,31 @@ int CrushWrapper::verify_upmap(CephContext *cct,
...
+            real_rep++;
+          }
+        }
+        if (real_rep != numrep) {
+          lderr(cct) << __func__ << " expected " << numrep << " items in bucket " << root_bucket
+                     << " real " << real_rep << dendl;
+          return -EINVAL;
+        }
+        // validate the osd's in subtree
+        for (int c = 0; cursor < up.size() && c < numrep; ++cursor, ++c) {
```
Comment (xiexingguo) [inline src/crush/CrushWrapper.cc]: hmm, this looks the wrong way to fix the problem to me.

actually I think you should instead verify that:
- we have exactly  2 up osds from datacenter-1
- we have exactly 2 up osds from datacenter-2
- there is no more remaining up osds...

Comment (hjwsm1989) [reply src/crush/CrushWrapper.cc]: - we have exactly 2 up (up[0] and up[1]) osds from datacenter-1

- we have exactly 2 up (up[2] and up[3]) osds from datacenter-2

- there is no more remaining up osds...

Comment (xiexingguo) [reply src/crush/CrushWrapper.cc]: ahh, i got you now

**Comment by xiexingguo** [comment]
can you fire a ticket to ensure this get backported?

**Comment by hjwsm1989** [comment]
https://tracker.ceph.com/issues/42485

**Pull request merged**